File: app/src/main/java/pl/laskowski/marcin/creating/SudokuShuffler.java
package pl.laskowski.marcin.creating;

import pl.laskowski.marcin.model.Field;
import pl.laskowski.marcin.model.Point;
import pl.laskowski.marcin.model.Sudoku;
import pl.laskowski.marcin.type.ClassicRectangle;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

/**
 * Created by Marcin Laskowski.
 */

public class SudokuShuffler {

    private final Random random;
    private final SudokuTransformations transformations = new SudokuTransformations();

    public SudokuShuffler() {
        random = new Random();
    }

    public SudokuShuffler(long seed) {
        random = new Random(seed);
    }

    public Sudoku shuffle(Sudoku sudoku, ClassicRectangle variant) {
        sudoku = rotateAndMirror(sudoku);
        sudoku = swapNumbers(sudoku, variant.regionSize());
        sudoku = swapColumns(sudoku, variant);
        sudoku = swapRows(sudoku, variant);
        return sudoku;
    }

    public Sudoku shuffleFull(Sudoku sudoku, ClassicRectangle variant){
        sudoku = swapColumns(sudoku, variant);
        sudoku = swapRows(sudoku, variant);
        return sudoku;
    }

    private Sudoku rotateAndMirror(Sudoku sudoku) {
        switch (random.nextInt(8)) {
            case 0:
                return sudoku.copy();
            case 1:
                return transformations.rotateLeft(sudoku);
            case 2:
                return transformations.rotate180(sudoku);
            case 3:
                return transformations.rotateRight(sudoku);
            case 4:
                return transformations.mirrorByXAxis(sudoku);
            case 5:
                return transformations.mirrorByYAxis(sudoku);
            case 6:
                return transformations.mirrorByFirstDiagonal(sudoku);
            case 7:
                return transformations.mirrorBySecondDiagonal(sudoku);
            default:
                throw new IllegalStateException();
        }
    }

    private Sudoku swapNumbers(Sudoku sudoku, int maxValue) {
        List<Integer> possibleValues = shuffledValues(maxValue);
        sudoku.forEach(field -> {
            if (!field.isEmpty()) {
                int currentValue = field.value();
                int newValue = possibleValues.get(currentValue - 1);
                field.set(newValue);
            }
        });
        return sudoku;
    }

    private Sudoku swapColumns(Sudoku sudoku, ClassicRectangle variant) {
        List<List<Field>> shuffled = shuffleGroups(sudoku.columns(), variant.getBlockWidth());
        return sudoku.copyWithIndexMapping(p -> {
            Field f = shuffled.get(p.x()).get(p.y());
            return new Point(f.getX(), f.getY());
        });
    }

    private Sudoku swapRows(Sudoku sudoku, ClassicRectangle variant) {
        List<List<Field>> shuffled = shuffleGroups(sudoku.rows(), variant.getBlockHeight());
        return sudoku.copyWithIndexMapping(p -> {
            Field f = shuffled.get(p.y()).get(p.x());
            return new Point(f.getX(), f.getY());
        });
    }

    private List<Integer> shuffledValues(int to) {
        List<Integer> values = new ArrayList<>();
        for (int value = 1; value <= to; value++) {
            values.add(value);
        }
        Collections.shuffle(values, random);
        return values;
    }

    private <T> List<T> shuffleGroups(List<T> origin, int size) {
        List<T> result = new ArrayList<>();
        List<List<T>> groups = new ArrayList<>();
        for (int i = 0; i < origin.size(); i += size) {
            groups.add(origin.subList(i, i + size));
        }
        Collections.shuffle(groups, random);
        for (List<T> group : groups) {
            Collections.shuffle(group, random);
            result.addAll(group);
        }
        return result;
    }

}

File: app/src/main/java/pl/laskowski/marcin/creating/SudokuShuffler.kt
package pl.laskowski.marcin.creating

import pl.laskowski.marcin.model.Field
import pl.laskowski.marcin.model.Sudoku
import pl.laskowski.marcin.model.type.SudokuType
import java.util.*

class SudokuShuffler(
    private val random: Random = Random()
) {
    private val transformations = SudokuTransformations()

    fun shuffle(sudoku: Sudoku, variant: SudokuType): Sudoku {
        return sudoku.rotateAndMirror()
            .swapNumbers(variant.regionSize)
            .swapColumns(variant)
            .swapRows(variant)
    }

    fun shuffleFull(sudoku: Sudoku, variant: SudokuType): Sudoku {
        return sudoku.swapColumns(variant)
            .swapRows(variant)
    }

    private fun Sudoku.rotateAndMirror(): Sudoku {
        return when (random.nextInt(8)) {
            0 -> copy()
            1 -> transformations.rotateLeft(this)
            2 -> transformations.rotate180(this)
            3 -> transformations.rotateRight(this)
            4 -> transformations.mirrorByXAxis(this)
            5 -> transformations.mirrorByYAxis(this)
            6 -> transformations.mirrorByFirstDiagonal(this)
            7 -> transformations.mirrorBySecondDiagonal(this)
            else -> throw IllegalStateException()
        }
    }

    private fun Sudoku.swapNumbers(maxValue: Int): Sudoku {
//        val possibleValues = shuffledValues(maxValue)
//        forEach(Consumer { field: Field? ->
//            if (field != null && !field.isEmpty) {
//                val currentValue = field.value()
//                val newValue = possibleValues[currentValue - 1]
//                field.set(newValue)
//            }
//        })
        return this
    }

    private fun Sudoku.swapColumns(variant: SudokuType): Sudoku {
//        val shuffled: List<List<Field>> = shuffleGroups(columns().assumeNoNulls(), variant.blockWidth)
//        return copyWithIndexMapping { (x, y): Point ->
//            val f = shuffled[x][y]
//            Point(f.x, f.y)
//        }
        return this
    }

    private fun Sudoku.swapRows(variant: SudokuType): Sudoku {
//        val shuffled: List<List<Field>> = shuffleGroups(rows().assumeNoNulls(), variant.blockHeight)
//        return copyWithIndexMapping { (x, y): Point ->
//            val f = shuffled[y][x]
//            Point(f.x, f.y)
//        }
        return this
    }

    private fun List<List<Field?>>.assumeNoNulls(): List<List<Field>> {
        this.flatten()
            .forEach { nullableField ->
                if (nullableField == null) {
                    throw IllegalStateException("Unexpected nulls")
                }
            }
        @Suppress("UNCHECKED_CAST")
        return this as List<List<Field>>
    }

    private fun shuffledValues(to: Int): List<Int> {
        val values: MutableList<Int> = ArrayList()
        for (value in 1..to) {
            values.add(value)
        }
        values.shuffle(random)
        return values
    }

    private fun <T> shuffleGroups(origin: List<T>, size: Int): List<T> {
        val result: MutableList<T> = ArrayList()
        val groups: MutableList<MutableList<T>> = ArrayList()
        var i = 0
        while (i < origin.size) {
            groups.add(origin.subList(i, i + size).toMutableList())
            i += size
        }
        groups.shuffle(random)
        for (group in groups) {
            group.shuffle(random)
            result.addAll(group)
        }
        return result
    }
}